File: iOS/BarcodeScanner/barcodescanner.js
/**
 * PhoneGap/Cordova is available under *either* the terms of the modified BSD license *or* the
 * MIT License (2008). See http://opensource.org/licenses/alphabetical for full text.
 *
 * Copyright (c) Matt Kane 2010
 * Copyright (c) 2010, IBM Corporation
 */

;(function(){

if (Cordova.hasResource("barcodeScanner")) return

Cordova.addResource("barcodeScanner")

//-------------------------------------------------------------------
var BarcodeScanner = function() {
}

//-------------------------------------------------------------------
BarcodeScanner.Encode = {
        TEXT_TYPE:     "TEXT_TYPE",
        EMAIL_TYPE:    "EMAIL_TYPE",
        PHONE_TYPE:    "PHONE_TYPE",
        SMS_TYPE:      "SMS_TYPE",
        CONTACT_TYPE:  "CONTACT_TYPE",
        LOCATION_TYPE: "LOCATION_TYPE"
}

//-------------------------------------------------------------------
BarcodeScanner.prototype.scan = function(success, fail, options) {
    function successWrapper(result) {
        result.cancelled = (result.cancelled == 1)
        success.call(null, result)
    }

    if (!fail) { fail = function() {}}

    if (typeof fail != "function")  {
        console.log("BarcodeScanner.scan failure: failure parameter not a function")
        return
    }

    if (typeof success != "function") {
        fail("success callback parameter must be a function")
        return
    }
  
    if ( null == options ) 
      options = []

    return Cordova.exec(successWrapper, fail, "org.apache.cordova.barcodeScanner", "scan", options)
}

//-------------------------------------------------------------------
BarcodeScanner.prototype.encode = function(type, data, success, fail, options) {
    if (!fail) { fail = function() {}}

    if (typeof fail != "function")  {
        console.log("BarcodeScanner.scan failure: failure parameter not a function")
        return
    }

    if (typeof success != "function") {
        fail("success callback parameter must be a function")
        return
    }

    return Cordova.exec(success, fail, "org.apache.cordova.barcodeScanner", "encode", [{type: type, data: data, options: options}])
}

//-------------------------------------------------------------------
Cordova.addConstructor(function() {
    if (!window.plugins) window.plugins = {}

    if (!window.plugins.barcodeScanner) {
        window.plugins.barcodeScanner = new BarcodeScanner()
    }
    else {
        console.log("Not installing barcodeScanner: window.plugins.barcodeScanner already exists")
    }
})

})();

File: BlackBerry/BarcodeScanner/BarcodeScanner.java
/*
 * PhoneGap is available under *either* the terms of the modified BSD license *or* the
 * MIT License (2008). See http://opensource.org/licenses/alphabetical for full text.
 *
 * Copyright (c) 2011, IBM Corporation
 */
package org.apache.cordova.plugins.barcodescanner;

import net.rim.device.api.ui.UiApplication;

import org.apache.cordova.api.Plugin;
import org.apache.cordova.api.PluginResult;
import org.apache.cordova.json4j.JSONArray;

/**
 * The BarcodeScanner plugin interface.
 *
 * The BarcodeScanner class can invoke the following actions:
 *
 *      scan: Launches a UI showing a live video feed from the camera and scans
 *            for barcodes.  If a barcode is detected and decoded the text is
 *            returned.
 *      encode: Encodes text data into a barcode image.
 */
public class BarcodeScanner extends Plugin {

    private static String ACTION_SCAN = "scan";
    private static String ACTION_ENCODE = "encode";
    static String TAG = "BarcodeScanner";

    private ScanAction scanner = null;
    private Thread scanThread = null;
    private Object lock = new Object();

    private UiApplication uiApp = UiApplication.getUiApplication();

    /**
     * Executes the requested action and returns a PluginResult.
     *
     * @param action
     *            The action to execute.
     * @param callbackId
     *            The callback ID to be invoked upon action completion.
     * @param args
     *            JSONArry of arguments for the action.
     * @return A PluginResult object with a status and message.
     */
    public PluginResult execute(String action, JSONArray args, String callbackId) {
        PluginResult result = null;

        if (ACTION_SCAN.equals(action)) {

            synchronized (lock) {
                // Only allow one active invocation of the scanner.
                if (scanner == null || !scanner.isDisplayed()) {
                    scanner = new ScanAction(callbackId, uiApp);

                    // Ensure that UI changes are done on event thread.
                    scanThread = new Thread(scanner);
                    scanThread.start();

                    result = new PluginResult(PluginResult.Status.NO_RESULT);
                    result.setKeepCallback(true);
                } else {
                    result = new PluginResult(PluginResult.Status.ERROR,
                            "Scanner already on display stack.");
                }
            }
        } else if (ACTION_ENCODE.equals(action)) {
            EncodeAction encoder = new EncodeAction();
            result = encoder.encode(args);
        } else {
            result = new PluginResult(PluginResult.Status.INVALID_ACTION,
                    TAG + ": Invalid action:" + action);
        }

        return result;
    }

    /**
     * Called when Plugin is paused.
     */
    public void onPause() {
        synchronized (lock) {
            if (scanner != null) {
                scanner.stop();
                scanner = null;
            }
        }
    }

}
